### PR TITLE
Create a custom Log manager for Kibana Dashboards

### PR DESCRIPTION
**Pull request by Levitan**
Issues #692 #1157

### CODE REVIEW

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+public class KibanaLogManager {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KibanaLogManager.class);
+
+    public void debug(String message, @Nullable Map<Object, Object> data) {
```
Comment (surabujin) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: em... what is this?

This "manager" must provide methods like "islDiscovery(SwitchId srcSwitch, int srcPort, SwitchId dstSwitch, int dstPort)". External world must know nothing about produced messages, all "logging" logic must be incapsulated inside `KibanaLogManager`.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+     * @param portId a port id.
+     * @param state a port state.
+     */
+    public void portDiscovery(SwitchId switchId, int portId, String state) {
+        Map<Object, Object> data =
+                ImmutableMap.of(TAG, "DISCO: Port Event", "switch", switchId, "port", portId, "state", state);
+        proceed("Port discovery", data, INFO);
+    }
+
+    public void portEvent(String state) {
```
Comment (surabujin) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: Where is the switchId and port number? How you going to report port event without any reference to port triggered this event?

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * ISL Discovery log.
+     *
+     * @param message a message.
+     * @param srcSwitch a source switch id.
+     * @param srcPort a source port id.
+     * @param dstPort a destination port id.
+     * @param state an ISL state.
+     */
+    public void islDiscovery(String message, SwitchId srcSwitch, Integer srcPort, Integer dstPort,
```
Comment (surabujin) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: Caller MUST have no control over produced log message. Why for we have "message" here?

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                             IslChangeType state) {
+        Map<Object, Object> data = ImmutableMap.of(TAG, "DISCO: ISL Event",
+                "srcSwitch", srcSwitch, "srcPort", srcPort, "dstPort", dstPort, "state", state);
+        proceed(message, data, INFO);
+    }
+
+    public void islEvent(IslChangeType state) {
```
Comment (surabujin) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: ... I am dropping review here - where is the detail about ISL? How this message can be used by person who read it?

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        }
+    }
+
+    /**
+     * Convert Map&lt;Object, Object&gt; to Map&lt;String, String&gt;
+     *
+     * @param data source MDC fields data
+     * @return a converted Map&lt;String, String&gt;
+     */
+    private Map<String, String> convertMap(Map<Object, Object> data) {
```
[the remark this answers is not in the document]
Comment (Levitan) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: With ImmutableMap, i collect in Map any Objects, this method convert Map<Object, Object> from ImmutableMap to Map<String, String> for correct logging

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(KibanaLogManager.class);
+
+    private static final String SWITCH = "switch";
+    private static final String SRC_SWITCH = "srcSwitch";
+    private static final String PORT = "port";
+    private static final String SRC_PORT = "srcPort";
+    private static final String DST_PORT = "dstPort";
+    private static final String STATE = "state";
+
+    private static final String MESSAGE_IGNORE_STATE = "Ignoring state";
```
[the remark this answers is not in the document]
Comment (Levitan) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: SonarQube write code smells about duplicate of String  "Ignoring state"

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+     *
+     * @param switchId a switch id.
+     * @param state a state of switch.
+     */
+    public void switchDiscovery(SwitchId switchId, SwitchState state) {
+        Map<String, String> data = new HashMap<>();
+        data.put(TAG, "DISCO: Switch Event");
+        data.put(SWITCH, switchId.toString());
+        data.put(STATE, state.toString());
+        proceed("Switch event", data);
```
[the remark this answers is not in the document]
Comment (Levitan) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: This case haven't info obout port number

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * The KIBANA dashboard log manager.
+ */
+@Slf4j
+public class KibanaLogManager {
+
+    private static final String SWITCH = "switch";
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: The constant name should indicate the purpose or scope or meaning of the constant. This one is unclear.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+import org.slf4j.MDC;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * The KIBANA dashboard log manager.
+ */
+@Slf4j
+public class KibanaLogManager {
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: Why manager? Can’t find what it manages, so It’s rather a logger or wrapper.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    private static final String TAG = "KIBANA_DASHBOARD_TAG";
+
+    /**
+     * Switch discovery dashboard log.
+     *
+     * @param switchId a switch id.
+     * @param state a state of switch.
+     */
+    public void switchDiscovery(SwitchId switchId, SwitchState state) {
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: I believe the method names should represent either some action or event. e.g. onSwitchDiscovery or logSwitchStateChange.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        proceed("State changed", data);
+    }
+
+    /**
+     * Build and write log message and MDC custom fields.
+     *
+     * @param message a message text.
+     * @param logData a data for MDC custom fields.
+     */
+    private void proceed(String message, Map<String, String> logData) {
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: I’d recommend to read the previous values and restore them on exit.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java (hunk 9 above)
Comment (Levitan) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: Rename to KibanaLogWrapper

File: services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java (hunk 10 above)
Comment (Levitan) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/utils/KibanaLogManager.java]: renamed

**Pull request merged**